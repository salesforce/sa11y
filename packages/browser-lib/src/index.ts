/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { exceptionListFilter, appendWcag } from '@sa11y/format';
import { defaultRuleset, registerCustomRules } from '@sa11y/preset-rules';
export { base, extended, full } from '@sa11y/preset-rules';
export const namespace = 'sa11y';

/**
 * Wrapper function to check accessibility to be invoked after the sa11y minified JS file
 * is injected into the browser
 * @param scope - Scope of the analysis as {@link A11yCheckableContext}
 * @param rules - preset sa11y rules defaulting to {@link base}
 * @param exceptionList - mapping of rule to css selectors to be filtered out using {@link exceptionListFilter}
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function checkAccessibility(
    scope = document,
    rules = defaultRuleset,
    exceptionList = {},
    addWcagInfo = false
) {
    // TODO (debug): adding type annotations to args, return type results in error:
    //  "[!] Error: Unexpected token" in both rollup-plugin-typescript2 and @rollup/plugin-typescript
    //  Compiling the index.ts file with tsc and using the dist/index.js file results
    //  in error when importing the "namespace" var. This would probably be easier to fix
    //  which could then result in getting rid of the rollup typescript plugins.

    // To register custom rules
    registerCustomRules();

    const results = await axe.run(scope || document, rules);
    const filteredResults = exceptionListFilter(results.violations, exceptionList);

    if (addWcagInfo) {
        appendWcag(filteredResults);
    }
    return JSON.stringify(filteredResults);
}
