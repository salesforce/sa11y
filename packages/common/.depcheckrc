# Ignore webdriverio - we need the types for `WebdriverIO.Browser`
ignores: ['webdriverio']
