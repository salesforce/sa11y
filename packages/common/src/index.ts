/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yConfig, AxeResults, axeRuntimeExceptionMsgPrefix, axeVersion, getAxeRules, getViolations } from './axe';
export { WdioAssertFunction, WdioOptions } from './wdio';
export { errMsgHeader, ExceptionList } from './format';
export { log, useFilesToBeExempted, useCustomRules, processFiles } from './helpers';
