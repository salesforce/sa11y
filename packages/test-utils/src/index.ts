/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export {
    audioURL,
    exceptionList,
    domWithA11yIssues,
    domWithDescendancyA11yIssues,
    domWithA11yIssuesBodyID,
    domWithNoA11yIssues,
    domWithNoA11yIssuesChildCount,
    domWithVisualA11yIssues,
    htmlFileWithA11yIssues,
    htmlFileWithNoA11yIssues,
    htmlFileWithVisualA11yIssues,
    a11yIssuesCount,
    a11yIssuesCountFiltered,
    shadowDomID,
    videoURL,
    customRulesFilePath,
    domWithA11yCustomIssues,
} from './test-data';
export { beforeEachSetup, cartesianProduct } from './utils';
