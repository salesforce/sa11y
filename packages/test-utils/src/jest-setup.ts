/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { extended, recommended } from '@sa11y/preset-rules';

// TODO: Can these functions be moved into @sa11y/jest as utils or as setupFiles so that API users can also use them?

/**
 * Common Jest setup that sets up jsdom and customizes preset-rules for jsdom
 */
export function beforeAllSetup(): void {
    document.documentElement.lang = 'en'; // required for a11y lang check

    // Customize rules specific to jsdom
    [extended, recommended].forEach(
        (config) =>
            (config.rules = {
                'color-contrast': { enabled: false }, // Disable color-contrast for jsdom
            })
    );
}

/**
 * Common Jest cleanup after each test
 */
export function afterEachCleanup(): void {
    document.body.innerHTML = ''; // reset dom body
}
