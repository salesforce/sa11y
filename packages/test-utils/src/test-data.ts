/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export const domWithA11yIssuesBodyID = 'dom-with-issues';
// DOM with no a11y issues
export const domWithA11yIssues = `<html>
                            <body id=${domWithA11yIssuesBodyID}>
                             <a href="#"></a>
                            </body>
                           </html>`;

export const shadowDomID = 'upside-down';
// DOM containing a11y issues
// From https://github.com/dequelabs/axe-selenium-java/blob/develop/src/test/resources/test-app.js-->
export const domWithNoA11yIssues = `<!doctype html>
                            <html lang="en">
                            <head>
                                <title>Test Page</title>
                            </head>
                            <body>
                            <div role="main" id="host">
                                <h1>This is a test</h1>
                                <p>This is a test page with no violations</p>
                            </div>
                            <div role="contentinfo" id=${shadowDomID}></div> <!-- cSpell:disable-line -->
                                <script>
                                    var shadow = document.getElementById("upside-down").attachShadow({mode: "open"});
                                    shadow.innerHTML = '<h2 id="shadow">SHADOW DOM</h2><ul><li>Shadow Item 1</li></ul>'
                                </script>
                            </body>
                            </html>`;
