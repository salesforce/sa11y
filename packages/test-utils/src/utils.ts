/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { axeRuntimeExceptionMsgPrefix } from '@sa11y/common';

/**
 * Cartesian product of arrays
 * Ref: https://eddmann.com/posts/cartesian-product-in-javascript/
 */
// TODO(types): Fix types for the cartesianProduct function - from any to generics
/* eslint-disable @typescript-eslint/no-explicit-any */
export function cartesianProduct(...sets: Array<any>): Array<any> {
    const flatten = (arr: Array<any>): Array<any> => [].concat([], ...arr);
    return sets.reduce((acc, set) => flatten(acc.map((x: any) => set.map((y: any) => [...x, y]))), [[]]);
    /* eslint-enable @typescript-eslint/no-explicit-any */
}

/**
 * Common Jest setup that sets up JSDOM as required for the tests
 */
export function beforeEachSetup(): void {
    document.documentElement.lang = 'en'; // required for a11y lang check
    document.body.innerHTML = ''; // reset body content
}

/**
 * Check if given error is an a11y error and not an axe runtime exception
 */
export function checkA11yError(e: Error): void {
    expect(e).toBeTruthy();
    expect(e.message).not.toContain(axeRuntimeExceptionMsgPrefix);
    expect(e.message).toMatchSnapshot();
}
