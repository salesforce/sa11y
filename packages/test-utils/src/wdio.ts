/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

// Ignore jest code coverage for WDIO test-utils
/* istanbul ignore file */

import { WdioAssertFunction, WdioOptions, axeRuntimeExceptionMsgPrefix, errMsgHeader } from '@sa11y/common';

export async function checkA11yErrorWdio(
    assertFunc: WdioAssertFunction,
    expectNumA11yIssues = 0,
    options: Partial<WdioOptions> = {}
): Promise<void> {
    // TODO (debug): setting expected number of assertions doesn't seem to be working correctly in mocha
    //  https://webdriver.io/docs/assertion.html
    //  Check mocha docs: https://mochajs.org/#assertions
    //  Checkout Jasmine ? https://webdriver.io/docs/frameworks.html
    // expect.assertions(99999); // still passes ???

    // TODO (debug): Not able to get the expect().toThrow() with async functions to work with wdio test runner
    //  hence using the longer try.. catch alternative
    // expect(async () => await assertAccessible()).toThrow();
    let err: Error = new Error();
    try {
        await assertFunc(options);
    } catch (e) {
        err = e as Error;
    }
    expect(err).toBeTruthy();
    expect(err.message).not.toContain(axeRuntimeExceptionMsgPrefix);

    if (expectNumA11yIssues > 0) {
        expect(err).not.toEqual(new Error());
        expect(err.toString()).toContain(`${expectNumA11yIssues} ${errMsgHeader}`);
    } else {
        expect(err).toEqual(new Error());
        expect(err.toString()).not.toContain(errMsgHeader);
    }
}
