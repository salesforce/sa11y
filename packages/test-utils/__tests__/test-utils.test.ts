/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { beforeEachSetup, cartesianProduct, checkA11yError } from '../src';
import { axeRuntimeExceptionMsgPrefix } from '@sa11y/common';
const testDOMCleanupContent = 'foo';

beforeAll(() => {
    // Populate DOM to test cleanup after each test
    document.body.innerHTML = testDOMCleanupContent;
});

beforeEach(() => {
    beforeEachSetup();
});

describe('test utils jest setup', () => {
    it('should setup jsdom as expected', () => {
        expect(document.documentElement.lang).toBe('en');
    });

    it('should cleanup document', () => {
        expect(document.body.innerHTML).not.toEqual(testDOMCleanupContent);
        expect(document.body.innerHTML).toEqual('');
    });
});

describe('test utils cartesian product', () => {
    const numArr = [0, 1, 2, 3, 4, 5];
    const alphabetArr = ['a', 'b', 'c', 'd', 'e'];
    it('should work as expected', () => {
        expect(cartesianProduct(numArr, alphabetArr)).toMatchSnapshot();
        expect(cartesianProduct(alphabetArr, numArr)).toMatchSnapshot();
        expect(cartesianProduct(numArr, numArr)).toMatchSnapshot();
        expect(cartesianProduct(alphabetArr, alphabetArr)).toMatchSnapshot();
    });
});

describe('test utils check a11y error', () => {
    it('should check for error to be defined', () => {
        expect.assertions(2);
        expect(() => checkA11yError(undefined)).toThrowErrorMatchingSnapshot();
    });

    it('should check for axe run time exception', () => {
        expect.assertions(3);
        expect(() => checkA11yError(new Error(axeRuntimeExceptionMsgPrefix))).toThrowErrorMatchingSnapshot();
    });

    it('should check for error to match snapshot', () => {
        expect.assertions(4);
        expect(() => checkA11yError(new Error('foo'))).not.toThrow();
    });
});
