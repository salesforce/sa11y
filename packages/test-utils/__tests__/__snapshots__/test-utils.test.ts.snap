// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test utils cartesian product should work as expected 1`] = `
Array [
  Array [
    0,
    "a",
  ],
  Array [
    0,
    "b",
  ],
  Array [
    0,
    "c",
  ],
  Array [
    0,
    "d",
  ],
  Array [
    0,
    "e",
  ],
  Array [
    1,
    "a",
  ],
  Array [
    1,
    "b",
  ],
  Array [
    1,
    "c",
  ],
  Array [
    1,
    "d",
  ],
  Array [
    1,
    "e",
  ],
  Array [
    2,
    "a",
  ],
  Array [
    2,
    "b",
  ],
  Array [
    2,
    "c",
  ],
  Array [
    2,
    "d",
  ],
  Array [
    2,
    "e",
  ],
  Array [
    3,
    "a",
  ],
  Array [
    3,
    "b",
  ],
  Array [
    3,
    "c",
  ],
  Array [
    3,
    "d",
  ],
  Array [
    3,
    "e",
  ],
  Array [
    4,
    "a",
  ],
  Array [
    4,
    "b",
  ],
  Array [
    4,
    "c",
  ],
  Array [
    4,
    "d",
  ],
  Array [
    4,
    "e",
  ],
  Array [
    5,
    "a",
  ],
  Array [
    5,
    "b",
  ],
  Array [
    5,
    "c",
  ],
  Array [
    5,
    "d",
  ],
  Array [
    5,
    "e",
  ],
]
`;

exports[`test utils cartesian product should work as expected 2`] = `
Array [
  Array [
    "a",
    0,
  ],
  Array [
    "a",
    1,
  ],
  Array [
    "a",
    2,
  ],
  Array [
    "a",
    3,
  ],
  Array [
    "a",
    4,
  ],
  Array [
    "a",
    5,
  ],
  Array [
    "b",
    0,
  ],
  Array [
    "b",
    1,
  ],
  Array [
    "b",
    2,
  ],
  Array [
    "b",
    3,
  ],
  Array [
    "b",
    4,
  ],
  Array [
    "b",
    5,
  ],
  Array [
    "c",
    0,
  ],
  Array [
    "c",
    1,
  ],
  Array [
    "c",
    2,
  ],
  Array [
    "c",
    3,
  ],
  Array [
    "c",
    4,
  ],
  Array [
    "c",
    5,
  ],
  Array [
    "d",
    0,
  ],
  Array [
    "d",
    1,
  ],
  Array [
    "d",
    2,
  ],
  Array [
    "d",
    3,
  ],
  Array [
    "d",
    4,
  ],
  Array [
    "d",
    5,
  ],
  Array [
    "e",
    0,
  ],
  Array [
    "e",
    1,
  ],
  Array [
    "e",
    2,
  ],
  Array [
    "e",
    3,
  ],
  Array [
    "e",
    4,
  ],
  Array [
    "e",
    5,
  ],
]
`;

exports[`test utils cartesian product should work as expected 3`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    0,
    1,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    3,
  ],
  Array [
    0,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
    0,
  ],
  Array [
    1,
    1,
  ],
  Array [
    1,
    2,
  ],
  Array [
    1,
    3,
  ],
  Array [
    1,
    4,
  ],
  Array [
    1,
    5,
  ],
  Array [
    2,
    0,
  ],
  Array [
    2,
    1,
  ],
  Array [
    2,
    2,
  ],
  Array [
    2,
    3,
  ],
  Array [
    2,
    4,
  ],
  Array [
    2,
    5,
  ],
  Array [
    3,
    0,
  ],
  Array [
    3,
    1,
  ],
  Array [
    3,
    2,
  ],
  Array [
    3,
    3,
  ],
  Array [
    3,
    4,
  ],
  Array [
    3,
    5,
  ],
  Array [
    4,
    0,
  ],
  Array [
    4,
    1,
  ],
  Array [
    4,
    2,
  ],
  Array [
    4,
    3,
  ],
  Array [
    4,
    4,
  ],
  Array [
    4,
    5,
  ],
  Array [
    5,
    0,
  ],
  Array [
    5,
    1,
  ],
  Array [
    5,
    2,
  ],
  Array [
    5,
    3,
  ],
  Array [
    5,
    4,
  ],
  Array [
    5,
    5,
  ],
]
`;

exports[`test utils cartesian product should work as expected 4`] = `
Array [
  Array [
    "a",
    "a",
  ],
  Array [
    "a",
    "b",
  ],
  Array [
    "a",
    "c",
  ],
  Array [
    "a",
    "d",
  ],
  Array [
    "a",
    "e",
  ],
  Array [
    "b",
    "a",
  ],
  Array [
    "b",
    "b",
  ],
  Array [
    "b",
    "c",
  ],
  Array [
    "b",
    "d",
  ],
  Array [
    "b",
    "e",
  ],
  Array [
    "c",
    "a",
  ],
  Array [
    "c",
    "b",
  ],
  Array [
    "c",
    "c",
  ],
  Array [
    "c",
    "d",
  ],
  Array [
    "c",
    "e",
  ],
  Array [
    "d",
    "a",
  ],
  Array [
    "d",
    "b",
  ],
  Array [
    "d",
    "c",
  ],
  Array [
    "d",
    "d",
  ],
  Array [
    "d",
    "e",
  ],
  Array [
    "e",
    "a",
  ],
  Array [
    "e",
    "b",
  ],
  Array [
    "e",
    "c",
  ],
  Array [
    "e",
    "d",
  ],
  Array [
    "e",
    "e",
  ],
]
`;
